LabKey/snprcEHRModules
# Epic 5642 & 5646: Updated timeline and animal selection (#97)

* Item 5642: SNPRC Scheduler Timeline Updates

- Project and timeline titles added to accordion panel titles
- Updated icons for procedure notes
- Add study day 0 field and handlers
- Add scheduled dates and handlers to grid
- Fix for procedure notes
- Formatting and functionality miscellaneous updates

* Item 5644: SNPRC Scheduler clone/revision, error handling, timeline updates

- Add clone, revision and delete functionality
- Confirm, alert and banner feedback
- Confirm before losing data on accordion navigation, timeline select and webpage navigation
- Update accordion navigation
- Day of the week column
- Study day note column
- Animal account field

* Item 5644: Scheduler timeline updates

- Cleanup reducers.
- Fix issue clearing study day 0 after save

* Item 5644: No permission message.  Project line list render issue

Only show banner with no permission message if 403 from server.  Fix render issue with project list rendering ouside of accordion panel.

* Epic 5642 & 5646: Updated timeline and animal selection

Updates to timeline based on user feedback.  Update accordion control to handle navigation to/from animal select.  Add animal reducer.  Add available/assigned grids with sorting, filtering and assigning functionality.  Fix bottom overflow issue with react boostrap table in available animal grid.  Handle end date in assigned animal grid

* Handle projects with no timelines. Add loading modal.

Fix bug when navigating to timeline view of project with no timelines.  Disable clone until timeline is created.  Disable revision and delete until timeline saved.  Cleanup router, now all hash pages redirect to app.view#/.  Cleanup ProjectList state to use redux and redux mapping.  Add loading mask when app is loading initial data, same mask as used for saving data.

* Fixed issue with timelines not in draft state generating an error (Ticket 37602) - added scheduler Admin and Reviewer roles to configuration page - fixed automated tests

* Epic 5642: Study day notes and package updates

- Update all packages to latest
- Add study day note handling
- Fix issue 37724: Save empty study days
- Fix issue 37725: New timelines are set to draft
- Fix issue with alert banner dismissing too early

* snprc_scheduler - fix for ticket 37794 - delete studyDayNotes in the UpdateTimelineItem code if needed

* Item 5645: Misc snprc scheduler fixes

- StudyDayNotes rowid issue causing row ids to disappear
- New timeline rowid issue causing clone/revised timelines not to display
- Clear state on deleting last timeline
- Fix delete draft issue
- Fix study day note clone/revision issue
- Fix repeated check/uncheck of last saved timeline item issue

* Fix test failure in FireFox

Issues with firefox are causing some issues loading from the APIs.  This is ultimately causing an issue when navigating away from the page in FF.

* Jest tests

Split animal reducer into new test suite.  Update tests.

* Set timelineItem IsDirty and fix NPE

- TimelineItem.IsDirty = true on update
- Fix NPE in animal reducer

* Recalculate row ids on save

Force recalculation of row ids on save.

* Fix NPE

Overzealous conditional application
